iceoryx_examples/waitset/README.md

```diff
@@ -1,6 +1,566 @@
-# waitset - Callbacks
+# WaitSet
 
-**Hint** This example shows a work in progress and is not yet working.
+The WaitSet is a set where you can attach Trigger to signal a wide variety
+of events to one single listener. The typical approach is that one creates a
+WaitSet attaches multiple subscriber or other Trigger to it and then wait till
+one or many of the attach classes signal an event. If that happens one receives
+a list of all the Triggers which were triggered and the program can act accordingly.
 
-## Introduction
+The WaitSet is state based which means that it will trigger until the state which
+caused the trigger changed. For instance, when the subscriber reports to the WaitSet
+that a new sample has been received, the WaitSet will notify the user until the
+sample has been taken and released.
 
+## Glossary
+
+ - **Listener** a class which can be triggered by a _Trigger_, for instance a _WaitSet_.
+ - **Trigger** a class which can be used to trigger a listener. If a _Trigger_ goes
```

Comment:
Why did you choose to rename both `Condition` to `Trigger` and `GuardConditon` to `UserTrigger`? The rmw interface is [heavily modelled](http://docs.ros2.org/beta2/api/rmw/rmw_8h.html#aece8df7e050626db62ff6fe7330cf7da) on DDS. IMHO we should rather stick to the DDS terminology for this reason. Also see rmw_iceoryx: https://github.com/ros2/rmw_iceoryx/blob/master/rmw_iceoryx_cpp/src/rmw_wait.cpp

Reply:
I would agree with you on that when we wouldn't have had redesigned it. I think the original DDS WaitSet design is great but also has some drawbacks which we tried to solve. The reason for renaming it was everyone asked me: Why is it called condition - it is a trigger. A trigger is triggered by a condition.

At the moment the idea is like the following:
1. A subscriber (which is the condition) acquires a Trigger by a WaitSet
2. If an event happens the subscriber will signal it (like the condition before) and triggers the Trigger
3. The WaitSet will ask via functor if the condition is true (which can be a bool method in a subscriber)

So in a way we still have conditions but they were removed since there is no need anymore for them since we solved the problem now with functors to bool method of to an arbitrary class instead of inheriting from the condition class and always calling `hasTriggered`.

Reply:
- For the DDS guys we could create some documentation on how the WaitSet differs from the DDS world. But in the end our WaitSet could be used nearly in the same way as the DDS WaitSet and offers even more use cases.
- The increase in memory is solvable. At the moment we have two interlinked triggers which are living in the WaitSet and the class attached to the WaitSet - the additional Trigger is the memory we pay. But if we could reduce it to one trigger and the other class just gets a pointer, we would solve the issue but for that we require a process local memory pool/allocator.